sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,390 @@
+using SIMD types because they need to use intrinsics or inline ASM, have legacy
+code that uses SIMD types, or are more familiar with that programming model.
+This extension adds support for an execution model in which SIMD parallelism is
+expressed using SIMD types, with clear delineation between SPMD and SIMD code regions through an invocation interface also introduced by the extension.
+
+=== Defining a SIMD Function
+
+Uniform arguments are represented by scalars, and varying arguments by explicit
+vector types.  The explicit vector types are based on the SIMD classes
+from [Parallelism TS2](http://wg21.link/N4808), and support all of the functions
+operating on SIMD types as defined there. Additionally, each of the functions
+defined in Sections 4.17.5, 4.17.6 and 4.17.7 of the SYCL 2020 specification
+are overloaded to accept SIMD classes; these overloads operate element-wise on
+the contents of the SIMD class.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+namespace simd_abi {
+template <class T, int N>
+using native_fixed_size = /* implementation-defined */;
+} // namespace simd_abi
+
+template <class T, int N>
+using simd = std::experimental::simd<T, simd_abi::native_fixed_size<N>>;
+
+template <class T, int N>
+using simd_mask = std::experimental::simd_mask<T, simd_abi::native_fixed_size<N>>;
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+We introduce a new `native_fixed_size` ABI which combines elements of the
+existing `native` and `fixed_size` ABIs defined in Parallelism TS2;
+implementations are expected to use the most efficient representation
+of a `simd<T, N>` type (e.g. using multiple registers rather than an array).
+We expect implementations of this ABI to build on Clang's `ext_vector_type`,
+rather than specifying the register types directly.
+
+The value of `N` for each `simd` argument to a SIMD function must be the same,
+and represents the sub-group size of the calling kernel.  If a function accepts
+`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
+kernel and the implementation must issue a diagnostic in this case.
+
+The example below shows a simple SIMD function that scales all elements of
+an 8-wide SIMD type `x` by a scalar value `n`.
+
+[source, c++]
+----
+sycl::ext::oneapi::simd<float, 8> scale(sycl::ext::oneapi::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+----
+
+=== Invoking a SIMD Function
+
+The `invoke_simd` free function invokes a SIMD function using all work-items
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
```

Comment:
(repeating the Q from internal discussion here)

Did you have some particular implementation approach in mind?

I’m thinking that invoke_simd will be an intrinsic which will capture the callable, arguments and argument/ret uniformity information (not 100% sure if it is needed). So that SPMD BE could do proper transformation and replace the call to invoke_simd with actual call after widening the caller.

If you had something similar in mind, then Ideally this would be a SPIRV intrinsic – is there a plan to standardize this part?

Reply:
I suspect that our implementation will look very similar to what you've described, at least for the backends that consume SPIR-V.  We need _something_ at the SPIR-V level to convey that we're transitioning from a SPMD context to a SIMD context, and a special `OpFunctionCall` seems like a good way to do that.  If we introduce such a thing, I think we should standardize it.

I don't think this specification should depend on it, though.  There are other valid implementations, and backends that don't go through SPIR-V should have the freedom to do something else.

Reply:
Good, thanks.

> I don't think this specification should depend on it, though.

Sure, agreed.